### PR TITLE
[tt-train] Lazy init

### PR DESCRIPTION
**Pull request by dyurshevichTT**
### Summary

Closes https://github.com/tenstorrent/tt-metal/issues/39804

Adds **deferred Python-side parameter initialization** for TTML models using `AbstractModuleBase` / `Parameter`, without empty autograd tensors as placeholders ([issue #39804](https://github.com/tenstorrent/tt-metal/issues/39804#issue-4071739670)).
### Implementation 

**Goal:** Build the Python module tree without allocating device weights or registering parameters in C++. One explicit step runs the same `ttml.init` logic and registers tensors so behavior matches eager construction (same seed / device / mappers).

**Phase 1 — declare (`with ttml.lazy_init():`)**

- A `ContextVar` flips lazy mode so `ttml.init.*` inner factories return **`TensorMetadata`** (shape + `init_fn` + optional `mapper`) instead of calling `ops.rand` / `from_numpy` / etc.
- **`Parameter(TensorMetadata(...))`** is stored on modules; **`AbstractModuleBase.__setattr__`** skips **`register_tensor`** until materialization.
- Accessing **`parameter.tensor`** (or in-place `ttml.init.*_` on a lazy parameter) raises with a clear message so forwards cannot run on placeholders.

**Phase 2 — realize (`ttml.materialize_module(model)`)**

- Walk the tree for **`Parameter`** slots still holding **`TensorMetadata`**.
- Materialize each distinct metadata object once (`id(metadata)`), assign the resulting autograd tensor into every **`Parameter`** that shared it.
- Resolve mapper: metadata mapper → optional **`layout_plan(full_path, meta)`** override → if multi-device mesh and still no mapper, use **all-replicate** `MeshMapperConfig` so non-sharded weights are not left single-device.
- **`_bind_parameter`** every `(module, attr)` that was lazy so **weight tying** registers the same tensor under each name (e.g. two modules, one shared **`Parameter`**).

**Init parity:** `init.py` uses **`_maybe_lazy`** plus shared **`_*_materialize`** helpers so eager and lazy call the same allocation code paths.

**Training:** `train_nanogpt.py` keeps order **open device → set seed → build**; with lazy enabled, build is **`lazy_init` + construct + `materialize_module` immediately after** so numerics align with eager.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:lazy-init)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:lazy-init)

### CODE REVIEW

**Comment by ayerofieiev-tt** [comment]
>This is not PyTorch meta tensors; it is Python-side deferral only—no empty autograd tensors as stand-ins.

Would you prefer pt style meta tensors instead?

**Comment by ayerofieiev-tt** [comment]
What happens with dynamic shapes in this lazy init?

File: tt-train/sources/ttml/nanobind/nb_core.cpp [hunk 1]
```diff
@@ -143,6 +143,12 @@ void py_module(nb::module_& m) {
             nb::arg("dim"),
             nb::arg("cluster_axis") = nb::none());
 
+        py_distributed.def(
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/nanobind/nb_core.cpp:146]: does this change belong here?

File: tt-train/sources/ttml/ttml/init.py [hunk 2]
```diff
@@ -273,16 +343,22 @@ def kaiming_uniform(
...
-        fan = _calculate_correct_fan(shape, mode)
-        gain = calculate_gain(nonlinearity, a)
-        std = gain / math.sqrt(fan)
-        bound = math.sqrt(3.0) * std
-        return uniform(-bound, bound)(shape)
+    def kaiming_uniform_init(shape, mapper=None):
+        def init_fn(s, m=None):
+            return _kaiming_uniform_materialize(s, a, mode, nonlinearity, m)
+
+        return _maybe_lazy(shape, init_fn, mapper)
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/ttml/init.py:350]: interestingly, i think this is how meta tensor works.
its one of the backends. there is a map between different ops impl in various backends and meta is one of the backends.

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by dyurshevichTT** [comment]
> > This is not PyTorch meta tensors; it is Python-side deferral only—no empty autograd tensors as stand-ins.
> 
> Would you prefer pt style meta tensors instead?

No, this line is super confusing, removed it :)

**Comment by dyurshevichTT** [comment]
> What happens with dynamic shapes in this lazy init?

This is only for weights initialization so no dynamic shapes.

File: tt-train/sources/ttml/nanobind/nb_core.cpp (hunk 1 above)
Comment (dyurshevichTT) [reply tt-train/sources/ttml/nanobind/nb_core.cpp:146]: Needed it as default mapper for multi device tensors for which user didn't specify a mapper.

File: tt-train/sources/ttml/ttml/modules/parameter.py [hunk 3]
```diff
@@ -4,17 +4,67 @@
...
+
+    shape: tuple[int, ...]
+    init_fn: Callable[..., Any]
+    mapper: Any | None = None
+    requires_grad: bool = True
+
+    def materialize(self, mapper_override: Any | None = None) -> Any:
+        """Allocate the autograd tensor using optional mapper override."""
+        mapper = self.mapper if mapper_override is None else mapper_override
+        return self.init_fn(self.shape, mapper)
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/modules/parameter.py:37]: I think copilot is correct here

File: tt-train/sources/ttml/ttml/modules/parameter.py [hunk 4]
```diff
@@ -4,17 +4,67 @@
...
+    """Deferred parameter: shape + factory; no device storage until materialization.
+
+    Do not use empty `ttml.autograd.Tensor` as a stand-in; keep allocation in
+    `init_fn` until `materialize()` runs.
+    """
+
+    shape: tuple[int, ...]
+    init_fn: Callable[..., Any]
+    mapper: Any | None = None
+    requires_grad: bool = True
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/modules/parameter.py:32]: Since we use this only for parameters anyway, maybe we should just remove this, and set requires grad to true always

File: tt-train/sources/ttml/ttml/lazy.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        if not isinstance(meta, TensorMetadata):
+            continue
+        mid = id(meta)
+        if mid in tensor_by_metadata_id:
+            tensor = tensor_by_metadata_id[mid]
+        else:
+            full_path = _full_path(mod, attr_name)
+            mapper = _resolve_mapper(full_path, meta)
+            tensor = meta.materialize(mapper_override=mapper)
+            tensor_by_metadata_id[mid] = tensor
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/lazy.py:121]: How can duplication happen? weight tying?

File: tt-train/tests/python/test_lazy_metadata_init.py
Comment (philei-tt) [inline tt-train/tests/python/test_lazy_metadata_init.py:1]: I think you don't test materialization with layout_plan and Sharding mapper?

File: tt-train/configs/training_configs/training_shakespeare_nanogpt_char.yaml [hunk 6]
```diff
@@ -15,9 +15,10 @@ training_config:
     beta1: 0.9
     beta2: 0.999
     epsilon: 1.0e-8
-    weight_decay: 0.01
-    amsgrad: false
-    stochastic_rounding: false
+  weight_decay: 0.01
+  amsgrad: false
+  stochastic_rounding: false
+  # lazy_parameter_init: false  # if true: defer Python Parameter alloc (train_nanogpt.py); default eager
```
Comment (imichalakTT) [inline tt-train/configs/training_configs/training_shakespeare_nanogpt_char.yaml:21]: Do we truly need to expose this option to configs?

File: tt-train/sources/ttml/ttml/modules/parameter.py [hunk 7]
```diff
@@ -4,17 +4,67 @@
...
+        mapper = self.mapper if mapper_override is None else mapper_override
+        return self.init_fn(self.shape, mapper)
 
 
 class Parameter:
     """Wrapper marking a tensor as a trainable parameter."""
 
     def __init__(self, tensor: Any) -> None:
-        self.tensor = tensor
+        object.__setattr__(self, "tensor", tensor)
```
Comment (imichalakTT) [inline tt-train/sources/ttml/ttml/modules/parameter.py:44]: I feel like it would be a lot easier to store the union type inside `Parameter._tensor` (maybe `Parameter.tensor_or_metadata`), and register `Parameter.tensor` as a property that throws for metadata, returns the tensor otherwise. Is there anything I miss?

File: tt-train/sources/ttml/ttml/init.py [hunk 8]
```diff
@@ -147,115 +148,184 @@ def calculate_gain(nonlinearity: _NonlinearityType, param: int | float | None =
...
+    """Host NumPy normal then device tensor (matches eager path on main)."""
+    data = np.random.normal(loc=mean, scale=std, size=tuple(shape)).astype(np.float32)
+    return ttml.autograd.Tensor.from_numpy(data, ttnn.Layout.TILE, ttnn.DataType.BFLOAT16, mapper)
+
+
 def normal(mean: float = 0.0, std: float = 1.0):
     """Normal (Gaussian) distribution."""
 
     def normal_init(shape, mapper=None):
-        data = np.random.normal(loc=mean, scale=std, size=tuple(shape)).astype(np.float32)
```
Comment (imichalakTT) [inline tt-train/sources/ttml/ttml/init.py:164]: Not a big deal, but instead of a third function definition we could make maybe_lazy a wrapper / decorator. Something like this:

```python
def maybe_lazy(init_fn):
    def wrapped(shape, mapper=None):
        shape = tuple(shape)

        if is_lazy_init_enabled():
            from ttml.modules.parameter import TensorMetadata
            return TensorMetadata(shape=shape, init_fn=init_fn, mapper=mapper)

        return init_fn(shape, mapper)

    return wrapped


def uniform(a: float = 0.0, b: float = 1.0):
    def uniform_init(shape, mapper=None):
        data = np.random.uniform(a, b, size=shape).astype(np.float32)

        return ttml.autograd.Tensor.from_numpy(
            data,
            ttnn.Layout.TILE,
            ttnn.DataType.BFLOAT16,
            mapper,
        )

    return maybe_lazy(uniform_init)
```

File: tt-train/sources/ttml/ttml/lazy.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            if m is mod:
+                return f"{prefix}.{attr_name}" if prefix else attr_name
+        return attr_name
+
+    def _resolve_mapper(full_path: str, meta: TensorMetadata) -> Any | None:
+        mapper = meta.mapper
+        if layout_plan is not None:
+            override = layout_plan(full_path, meta)
+            if override is not None:
+                mapper = override
```
Comment (imichalakTT) [inline tt-train/sources/ttml/ttml/lazy.py:102]: I prefer more concise form, but this works

```python
if layout_plan is not None and (m := layout_plan(full_path, meta)) is not None:
    mapper = m
```

**Review by imichalakTT — APPROVED** [review]
Please address Copilot comments

**Comment by imichalakTT** [comment]
Could you provide more context for why we need lazy init?

The summary from the issue is not really applicable anymore to the TP because we went with an explicit TP initialization where sharding is known during allocation from Mesh.

> Refactor Python module initialization to defer tensor allocation until sharding layout is known, eliminating wasteful full-tensor allocation followed by gather-shard-reallocate cycle.

File: tt-train/sources/ttml/ttml/modules/parameter.py [hunk 10]
```diff
@@ -4,17 +4,67 @@
...
+        """Allocate the autograd tensor using optional mapper override."""
+        mapper = self.mapper if mapper_override is None else mapper_override
+        return self.init_fn(self.shape, mapper)
 
 
 class Parameter:
     """Wrapper marking a tensor as a trainable parameter."""
 
     def __init__(self, tensor: Any) -> None:
-        self.tensor = tensor
```
Comment (awliu-TT) [inline tt-train/sources/ttml/ttml/modules/parameter.py:14]: Seems like `def __setattr__` just calls `object.__setattr__`. Do we need the explicit `object.__setattr__(self, "tensor", tensor)` here then?

**Review by awliu-TT — APPROVED** [review]
